### PR TITLE
Comparison for train between torch.sigmoid and fast_sigmoid

### PR DESCRIPTION
**Pull request by hieplcc**
**Comparison for train between torch.sigmoid and fast_sigmoid:** 
`/extension-cpp/train/train_fast_sigmoid.py`
Outputs:
```
============================================================
TRAINING WITH TORCH.SIGMOID (ORIGINAL)
============================================================
Epoch  10 | Loss: 1.0659 | Train Acc: 0.6550
Epoch  20 | Loss: 0.9921 | Train Acc: 0.8125
Epoch  30 | Loss: 0.8563 | Train Acc: 0.8187
Epoch  40 | Loss: 0.6952 | Train Acc: 0.8050
Epoch  50 | Loss: 0.5656 | Train Acc: 0.8225
Epoch  60 | Loss: 0.4675 | Train Acc: 0.8538
Epoch  70 | Loss: 0.3901 | Train Acc: 0.8650
Epoch  80 | Loss: 0.3366 | Train Acc: 0.8775
Epoch  90 | Loss: 0.3054 | Train Acc: 0.8862
Epoch 100 | Loss: 0.2874 | Train Acc: 0.8838
Test Accuracy: 0.8700

============================================================
TRAINING WITH FAST_SIGMOID (REPLACEMENT)
============================================================
Epoch  10 | Loss: 1.0787 | Train Acc: 0.3450
Epoch  20 | Loss: 1.0319 | Train Acc: 0.7238
Epoch  30 | Loss: 0.9366 | Train Acc: 0.8462
Epoch  40 | Loss: 0.7761 | Train Acc: 0.8650
Epoch  50 | Loss: 0.6019 | Train Acc: 0.8662
Epoch  60 | Loss: 0.4702 | Train Acc: 0.8688
Epoch  70 | Loss: 0.3820 | Train Acc: 0.8762
Epoch  80 | Loss: 0.3268 | Train Acc: 0.8850
Epoch  90 | Loss: 0.2953 | Train Acc: 0.8825
Epoch 100 | Loss: 0.2773 | Train Acc: 0.8812
Test Accuracy: 0.8650

============================================================
CONVERGENCE COMPARISON RESULTS
============================================================
Metric               torch.sigmoid   fast_sigmoid    Difference     
-----------------------------------------------------------------
Final Loss           0.287414        0.277347        0.010067       
Test Accuracy        0.8700          0.8650          0.0050         
Loss Diff %          -               -               3.50   
```

### CODE REVIEW

File: train/train_fast_sigmoid.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        self.fc3 = nn.Linear(8, 3)
+
+    def forward(self, x):
+        x = fast_sigmoid(self.fc1(x), -5, 5, 2000)  # Replace torch.sigmoid with fast_sigmoid
+        x = fast_sigmoid(self.fc2(x), -5, 5, 2000)  # Replace torch.sigmoid with fast_sigmoid
+        x = self.fc3(x)
+        return x
+
+# Train fast_sigmoid model with identical setup
+model_fast = FastSigmoidMLP()
```
[the remark this answers is not in the document]
Comment (loctxmoreh) [reply train/train_fast_sigmoid.py]: This copilot comment is relevant. You can follow this suggestion

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        message(FATAL_ERROR "Could not find PyTorch installation. Make sure PyTorch is installed and accessible via python3.")
+    endif()
+endif()
+
+# Find PyTorch
+find_package(Torch REQUIRED)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${TORCH_CXX_FLAGS}")
+
+# Add include directories
+include_directories(extension_cpp/csrc/include)
```
[the remark this answers is not in the document]
Comment (loctxmoreh) [reply CMakeLists.txt:29]: I think this copilot comment is fairly correct

File: train/train_fast_sigmoid.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+class FastSigmoidMLP(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.fc1 = nn.Linear(2, 16)
+        self.fc2 = nn.Linear(16, 8)
+        self.fc3 = nn.Linear(8, 3)
+
+    def forward(self, x):
+        x = fast_sigmoid(self.fc1(x), -5, 5, 2000)  # Replace torch.sigmoid with fast_sigmoid
+        x = fast_sigmoid(self.fc2(x), -5, 5, 2000)  # Replace torch.sigmoid with fast_sigmoid
```
Comment (loctxmoreh) [inline train/train_fast_sigmoid.py]: Can you explain the choice of hyper params here? What would happen if you just use the default arguments `(-10.0, 10.0, 1000)` ?

File: train/train_fast_sigmoid.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+acc_diff = abs(torch_test_acc - fast_test_acc)
+loss_diff_percent = loss_diff / torch_final_loss * 100
+
+print(f"{'Metric':<20} {'torch.sigmoid':<15} {'fast_sigmoid':<15} {'Difference':<15}")
+print("-" * 65)
+print(f"{'Final Loss':<20} {torch_final_loss:<15.6f} {fast_final_loss:<15.6f} {loss_diff:<15.6f}")
+print(f"{'Test Accuracy':<20} {torch_test_acc:<15.4f} {fast_test_acc:<15.4f} {acc_diff:<15.4f}")
+print(f"{'Loss Diff %':<20} {'-':<15} {'-':<15} {loss_diff_percent:<15.2f}%")
+
+# How to assess convergence?
```
Comment (loctxmoreh) [inline train/train_fast_sigmoid.py:170]: Suggestion: beside plotting the decision boundary, you can plot and compare the two loss curves

**Review by loctxmoreh — COMMENTED** [review]
Minor comments only

File: train/train_fast_sigmoid.py (hunk 3 above)
Comment (hieplcc) [reply train/train_fast_sigmoid.py]: I Just tried different argument sets to tune the accuracy. The reasons I choose (-5, 5, 2000) here are because:
- More entries (2000 entries) and tighter range, higher interpolation accuracy I have.
- And I know that sigmoid(-5) = 0.0067 (nearlt 0) and sigmoid(5) = 0.9933 (nearly 1), so outside this range, it's nearly flat.

For the default arguments, I will compare using them and give you the result later. But I think, in generally, (-5, 5, 2000) would get a better result.

Comment (loctxmoreh) [reply train/train_fast_sigmoid.py]: I've tried it myself and there seems to be no difference. 
Generally, i think the default arguments should be good enough for convergence.

**Review by loctxmoreh — APPROVED** [review]
The training script converges, so this PR LGTM 

Optionally, you can plot the two loss curves (one with `torch.sigmoid` and one with `fast_sigmoid`) and the difference should be minimal

File: train/train_fast_sigmoid.py (hunk 4 above)
Comment (hieplcc) [reply train/train_fast_sigmoid.py:170]: Dear @loctxmoreh 
I just pushed the commit to plot and compare the two loss curves.
The commit also removed duplicate codes suggested by Copilot.
Please help me to review https://github.com/hieplcc/extension-cpp/pull/2/commits/8f0cd9feaea1555508524678e6ccfa79a943eff6

**Review by loctxmoreh — APPROVED** [review]
I've finished my review on this PR. LGTM
